crates/red_knot_python_semantic/src/types/mro.rs

```diff
@@ -0,0 +1,401 @@
+
+/// The inferred method resolution order of a given class.
+///
+/// See [`ClassType::mro`] for more details.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub(super) struct Mro<'db>(Box<[ClassBase<'db>]>);
+
+impl<'db> Mro<'db> {
+    /// In the event that a possible list of bases would (or could) lead to a
+    /// `TypeError` being raised at runtime due to an unresolvable MRO, we
+    /// infer the class as being `[<the class in question>, Unknown, object]`.
+    /// This seems most likely to reduce the possibility of cascading errors
+    /// elsewhere.
+    ///
+    /// (We emit a diagnostic warning about the runtime `TypeError` in
+    /// [`super::infer::TypeInferenceBuilder::infer_region_scope`].)
+    pub(super) fn from_error(db: &'db dyn Db, class: ClassType<'db>) -> Self {
+        Self::from([
+            ClassBase::Class(class),
+            ClassBase::Unknown,
+            ClassBase::object(db),
+        ])
+    }
+
+    /// Attempt to resolve the MRO of a given class
+    pub(super) fn of_class(db: &'db dyn Db, class: ClassType<'db>) -> Result<Self, MroError<'db>> {
+        let class_stmt_node = class.node(db);
+
+        match class_stmt_node.bases() {
+            // `builtins.object` is the special case:
+            // the only class in Python that has an MRO with length <2
+            [] if class.is_known(db, KnownClass::Object) => {
+                Ok(Self::from([ClassBase::Class(class)]))
+            }
+
+            // All other classes in Python have an MRO with length >=2.
+            // Even if a class has no explicit base classes,
+            // it will implicitly inherit from `object` at runtime;
+            // `object` will appear in the class's `__bases__` list and `__mro__`:
+            //
+            // ```pycon
+            // >>> class Foo: ...
+            // ...
+            // >>> Foo.__bases__
+            // (<class 'object'>,)
+            // >>> Foo.__mro__
+            // (<class '__main__.Foo'>, <class 'object'>)
+            // ```
+            [] => Ok(Self::from([ClassBase::Class(class), ClassBase::object(db)])),
+
+            // Fast path for a class that has only a single explicit base.
+            //
+            // This *could* theoretically be handled by the final branch below,
+            // but it's a common case (i.e., worth optimizing for),
+            // and the `c3_merge` function requires lots of allocations.
+            [single_base_node] => {
+                let single_base = ClassBase::try_from_node(
+                    db,
+                    single_base_node,
+                    class_stmt_node,
+                    class.definition(db),
+                );
+                match single_base {
+                    Ok(base) => {
+                        if base.into_class_literal_type() == Some(class) {
+                            Err(MroError::CyclicClassDefinition {
+                                invalid_base_index: 0,
+                            })
```

Comment:
This approach won't work for handling indirect cycles. I think you'll need to test if we appear anywhere in `Mro::of_base().elements()`?

Reply:
> This approach won't work for handling indirect cycles. I think you'll need to test if we appear anywhere in `Mro::of_base().elements()`?

Hmm, I tried that and it doesn't fix the tests that @MichaReiser suggested in https://github.com/astral-sh/ruff/pull/14027#discussion_r1825877438; we still panic on those. I think I need to do something a bit fancier (I have an idea).

The diff I tried was this:

<details>

```diff
--- a/crates/red_knot_python_semantic/src/types/mro.rs
+++ b/crates/red_knot_python_semantic/src/types/mro.rs
@@ -79,10 +79,19 @@ impl<'db> Mro<'db> {
                             };
                             Err(MroError::new(db, class, error_kind))
                         } else {
-                            let mro = std::iter::once(ClassBase::Class(class))
-                                .chain(Mro::of_base(db, base).elements())
-                                .collect();
-                            Ok(mro)
+                            let class_as_base = ClassBase::Class(class);
+                            let base_mro = Mro::of_base(db, base);
+                            if base_mro.contains(&class_as_base) {
+                                let error_kind = MroErrorKind::CyclicClassDefinition {
+                                    invalid_base_index: 0,
+                                };
+                                Err(MroError::new(db, class, error_kind))
+                            } else {
+                                let mro = std::iter::once(class_as_base)
+                                    .chain(base_mro.elements())
+                                    .collect();
+                                Ok(mro)
+                            }
                         }
                     }
                     Err(invalid_base_ty) => {
@@ -107,11 +116,21 @@ impl<'db> Mro<'db> {
                     match ClassBase::try_from_node(db, base_node, class_stmt_node, definition) {
                         Ok(valid_base) => {
                             if valid_base.into_class_literal_type() == Some(class) {
+                                let error_kind = MroErrorKind::CyclicClassDefinition {
+                                    invalid_base_index: 0,
+                                };
+                                return Err(MroError::new(db, class, error_kind));
+                            }
+
+                            let class_as_base = ClassBase::Class(class);
+                            let base_mro = Mro::of_base(db, valid_base);
+                            if base_mro.contains(&class_as_base) {
                                 let error_kind = MroErrorKind::CyclicClassDefinition {
                                     invalid_base_index: i,
                                 };
                                 return Err(MroError::new(db, class, error_kind));
                             }
+
                             valid_bases.push(valid_base);
```

</details>

Reply:
Oh, yeah, because trying to calculate the MRO of the base gets us right back to here, and we get into the infinite cycle before ever checking for our presence in the base MRO.

Looking forward to the fancier version! I suspect you'll have to keep a set of seen classes and pass it down, or similar.

Reply:
> Looking forward to the fancier version! I suspect you'll have to keep a set of seen classes and pass it down, or similar.

yeah...